lib/pushmi_pullyu/aip/downloader.rb

```diff
@@ -1,215 +1,184 @@
   end
 
-  def base_path
+  def aip_directory
     PushmiPullyu::AIP.aip_directory(@noid)
   end
 
-  def data_path
-    "#{base_path}/data"
-  end
-
-  def full_local_path(path)
-    "#{data_path}/#{path}"
-  end
-
   ### Files
 
   def aip_paths
     @aip_paths ||= OpenStruct.new(
       main_object: OpenStruct.new(
         remote: nil, # Base path
-        local: "#{@aip_dirs.metadata}/object_metadata.n3",
+        local: "#{aip_dirs.metadata}/object_metadata.n3",
         optional: false
       ),
       fixity: OpenStruct.new(
         remote: '/content/fcr:fixity',
-        local: "#{@aip_dirs.logs}/content_fixity_report.n3",
+        local: "#{aip_dirs.logs}/content_fixity_report.n3",
         optional: false
       ),
       content_datastream_metadata: OpenStruct.new(
         remote: '/content/fcr:metadata',
-        local: "#{@aip_dirs.metadata}/content_fcr_metadata.n3",
+        local: "#{aip_dirs.metadata}/content_fcr_metadata.n3",
         optional: false
       ),
       versions: OpenStruct.new(
         remote: '/content/fcr:versions',
-        local: "#{@aip_dirs.metadata}/content_versions.n3",
-        optional: false
-      ),
-      content: OpenStruct.new(
-        remote: '/content',
-        local: nil, # Filename derived from metadata
+        local: "#{aip_dirs.metadata}/content_versions.n3",
         optional: false
       ),
 
       # Optional downloads
       thumbnail: OpenStruct.new(
         remote: '/thumbnail',
-        local: "#{@aip_dirs.thumbnails}/thumbnail",
+        local: "#{aip_dirs.thumbnails}/thumbnail",
         optional: true
       ),
       characterization: OpenStruct.new(
         remote: '/characterization',
-        local: "#{@aip_dirs.logs}/content_characterization.n3",
+        local: "#{aip_dirs.logs}/content_characterization.n3",
         optional: true
       ),
       fedora3foxml: OpenStruct.new(
         remote: '/fedora3foxml',
-        local: "#{@aip_dirs.metadata}/fedora3foxml.xml",
+        local: "#{aip_dirs.metadata}/fedora3foxml.xml",
         optional: true
       ),
       fedora3foxml_metadata: OpenStruct.new(
         remote: '/fedora3foxml/fcr:metadata',
-        local: "#{@aip_dirs.metadata}/fedora3foxml.n3",
+        local: "#{aip_dirs.metadata}/fedora3foxml.n3",
         optional: true
       )
     ).freeze
   end
 
+  # Extract filename from main object metadata
   def content_filename
-    return @content_filename unless @content_filename.nil?
-
-    # Extract filename from main object metadata
-    graph = RDF::Graph.load(full_local_path(aip_paths.main_object.local))
-    graph.each_statement do |statement|
-      if statement.predicate == FILENAME_PREDICATE
-        @content_filename = "#{aip_dirs.objects}/#{statement.object}"
-        break
-      end
+    filename_predicate = RDF::URI('info:fedora/fedora-system:def/model#downloadFilename')
+
+    graph = RDF::Graph.load(aip_paths.main_object.local)
+
+    graph.query(predicate: filename_predicate) do |results|
```

Comment:
use RDF query instead of looping through every node of the graph. More efficient and cleaner.